rfcs/rdm-0068-data-migration.md

```diff
@@ -0,0 +1,369 @@
+        """Returns the update date of the record."""
+        pass
+
+    # all functions used in the transform method are defined
+    # as abstract methods, they have been removed from simplicity.
+
+    def transform(self, entry):
+        """Transform a record single entry."""
+        return {
+            "created": self._created(entry),
+            "updated": self._updated(entry),
+            "version_id": self._version_id(entry),
+            "index": self._index(entry),
+            "json": {
+                "id": self._recid(entry),
+                "pids": self._pids(entry),
+                "files": self._files(entry),
+                "metadata": self._metadata(entry),
+                "access": self._access(entry),
+            },
+        }
+```
+
+At this point, the million dollar question, is _Why not Marshmallow?_. On the
+first place, Marshmallow is slow in processing, and performance is one of the
+key aspects to optimize in the migration. Then, most of the data transformation
+is custom, so the resulting Marshmallow schemas would end up being purely
+`Method` fields. Finally, validating (mostly referential, like vocabularies)
+can only be done on/after loading the data to the new instance where there is
+access to the database and application context.
+
+#### Load
+
+Loading the data into the new InvenioRDM instance is the last step of the
+stream. Data needs to be validated, prepared and inserted into the corresponding
+tables. The preparation step, enables optimizations such as bulk inserts or 
+loading from files.
+
+```python
+class Load(ABC):
+    """Base class for data loading."""
+
+    @abstractmethod
+    def _validate(self):
+        """Validate data before loading."""
+        pass
+
+    @abstractmethod
+    def _prepare(self):
+        """Prepare data for loading."""
+        pass
+
+    @abstractmethod
+    def _load(self):
+        """load data."""
+        pass
+
+    def run(self, entries):
+        """Load entries."""
+        for entry in entries:
+            if self._validate(entry):
+                self._prepare(entry)
+                self._load(entry)
+```
+
+The current implementation of `PostgreSQLCopyLoad` makes user of one or more
+`TableGenerator`. These generators create a csv file per table (they can output
+more than one table, e.g. records and drafts), which then will be used to `COPY`
+in bulk in the `_load` call being hundreds of times more performant than
+executing one `INSERT` per row.
+
+#### Shared Cache
+
+Streams might have interdependencies. Therefore during a migration run, there is
+a need to pass information from one stream to another. For example, the records
+stream need information about the community ids and slugs, which comes from the
+communities stream. A similar case can happens between `TableGenerators`, for
+example the records stream must keep track of the "seen" parent ids so it can
+update the parent information for different record versions. 
```

Comment:
I am not sure I understand the last sentence. Maybe "seen" -> reserved, existing ?

> so it can update the parent information for different record versions

nit: is it updating or inserting the parent info from scratch ?

Reply:
rephrased